cincinnati/src/plugins/internal/cincinnati_graph_fetch.rs

```diff
@@ -0,0 +1,308 @@
+            "http_upstream_requests_total",
+            "Total number of HTTP upstream requests",
+        )?;
+
+        let http_upstream_errors_total = Counter::new(
+            "http_upstream_errors_total",
+            "Total number of HTTP upstream unreachable errors",
+        )?;
+
+        if let Some(registry) = &prometheus_registry {
+            registry.register(Box::new(http_upstream_reqs.clone()))?;
+            registry.register(Box::new(http_upstream_errors_total.clone()))?;
+        };
+
+        Ok(Self {
+            upstream,
+            http_upstream_reqs,
+            http_upstream_errors_total,
+        })
+    }
+}
+
+impl InternalPlugin for CincinnatiGraphFetchPlugin {
+    fn run_internal(self: &Self, io: InternalIO) -> AsyncIO<InternalIO> {
+        use crate::CONTENT_TYPE;
+        use actix_web::http::header::{self, HeaderValue};
+        use commons::GraphError;
+        use futures::{future, Future, Stream};
+        use hyper::{Body, Client, Request};
+
+        let upstream = self.upstream.to_owned();
+
+        trace!("getting graph from upstream at {}", upstream);
+
+        // Assemble a request for the upstream Cincinnati service.
+        let ups_req = match Request::get(upstream)
+            .header(header::ACCEPT, HeaderValue::from_static(CONTENT_TYPE))
+            .body(Body::empty())
+        {
+            Ok(req) => req,
+            Err(e) => {
+                error!("failed to create upstream request: {}", e);
+                return Box::new(future::err(failure::Error::from(
+                    GraphError::FailedUpstreamRequest(e.to_string()),
+                )));
+            }
+        };
+
+        self.http_upstream_reqs.inc();
+
+        let http_upstream_errors_total_failed_request = self.http_upstream_errors_total.clone();
+        let http_upstream_errors_total_wrong_status = self.http_upstream_errors_total.clone();
+
+        let future_graph = Client::new()
+            .request(ups_req)
+            .map_err(move |e| {
+                error!("failed upstream request. {}", e);
+                http_upstream_errors_total_failed_request.inc();
+                GraphError::FailedUpstreamFetch(e.to_string())
+            })
+            .and_then(move |res| {
+                if res.status().is_success() {
+                    future::ok(res)
+                } else {
+                    error!("unsucessful status on upstream request: {}", res.status());
+                    http_upstream_errors_total_wrong_status.inc();
+                    future::err(GraphError::FailedUpstreamFetch(res.status().to_string()))
+                }
+            })
+            .and_then(|res| {
+                res.into_body()
+                    .concat2()
+                    .map_err(|e| GraphError::FailedUpstreamFetch(e.to_string()))
+            })
+            .and_then(|body| {
+                serde_json::from_slice(&body).map_err(|e| GraphError::FailedJsonIn(e.to_string()))
```

Comment:
It looks like this error case is not accounted into the metrics counter.

Reply:
The bottom `.map_err()` will also catch this one